pycarol/bigquery.py

```diff
@@ -140,6 +142,118 @@ def query(
         else:
             return results
 
+    def paginated_query(
+        self,
+        query: str,
+        page_size=500, 
+        dataset_id: T.Optional[str] = None,
+        return_dataframe: bool = True,
+    ):
+        """Run query for datamodel and prepare paginated results. This will generate a SA if necessary.
+
+        Args:
+            query: BigQuery SQL query.
+            page_size: The number of records per page. If not provided defaults to 500.
+            dataset_id: BigQuery dataset ID.
+                if None it will use the default dataset_id.
+            return_dataframe: Return dataframe.
+                Return dataframe if True.
+
+        Returns:
+            Query results for the first page and control variables as a tuple: (results, control_dict)
+
+        Usage:
+
+        .. code:: python
+
+
+            from pycarol import Carol
+            from pycarol.bigquery import BQ
+
+            bq = BQ(Carol())
+            query = 'select * from invoice limit 10'	
+            df_first_page, c = bq.paginated_query(query, page_size=2, return_dataframe=True)
+
+        """
+
+        self.service_account = self.get_credential()
+        self.client = self._generate_client()
+
+        dataset_id = dataset_id or self.dataset_id
+        job_config = bigquery.QueryJobConfig(default_dataset=dataset_id)
+        job = self.client.query(query, job_config=job_config)
```

Comment:
```python
        self.service_account = self.get_credential()
        self.client = self._generate_client()

        dataset_id = dataset_id or self.dataset_id
        job_config = bigquery.QueryJobConfig(default_dataset=dataset_id)
        job = self.client.query(query, job_config=job_config)
```

is the same here and the maiun `self.query` 

I suggest having a 
```python
self._query_job(self, fetch_page):
             self.service_account = self.get_credential()
        self.client = self._generate_client()

        dataset_id = dataset_id or self.dataset_id
        job_config = bigquery.QueryJobConfig(default_dataset=dataset_id)
        job = self.client.query(query, job_config=job_config)
        return job
```
What do you think?

Reply:
We can use `._query_job` to get the job on `query` and `paginated_query` methods. For `fetch_page` we retrieve it from the Job ID. 
I think it may be worth creating a class for this job, so that we can store the page_size on it and avoid confusion with user calling `fetch_page` with different page sizes...

Reply:
yes yes, I meant fot those two `query` and `paginated_query`